### PR TITLE
Save deduplicated traditional XBRL instance

### LINKED ISSUE #1002: Add Option to Deduplicate Facts from Traditional XBRL
### What should we change and why?

We should reuse the implementation in #1001 to offer the ability to deduplicate complete or consistent facts from traditional XBRL instance documents.

```bash
python arelleCmdLine.py --file instance.xml --deduplicateFacts=consistent --saveDeduplicateInstance=report-deduplicated.xml
```

`--deduplicateFacts` should default to `complete` and an exception should be raised if `--deduplicateFacts` is used without `--saveDeduplicateInstance`.

* [ ] complete duplicates.
* [ ] consistent duplicates (the more precise fact should be preserved).

### PR DESCRIPTION
**Pull request by aaroncameron-wk**
#### Reason for change
Resolves #1002 

#### Description of change
Reused parts of duplicate fact validation to transform duplicate fact sets into a deduplicated list based on the provided argument. Implemented `complete`, `consistent-set`, and `consistent-pairs` deduplication modes.
Added duplicate fact user guide to documentation.
Added fact deduplication integration test script.

#### Steps to Test
CI

**review**:
@Arelle/arelle

### CODE REVIEW

**Comment by aviary2-wf** [comment]
## Security Insights
No security relevant content was detected by automated scans.

## Action Items
* Review PR for [security impact](https://wiki.atl.workiva.net/display/SECURITY/Development+Security+Review+Guidelines); comment "security review required" if needed or unsure
* Verify [`aviary.yaml`](https://wiki.atl.workiva.net/pages/viewpage.action?pageId=42762752#SecurityInsightFramework&Tooling-ConfiguringAviary) coverage of security relevant code
---
_Questions or Comments? Reach out on Slack: #support-infosec._

File: arelle/CntlrCmdLine.py
Comment (austinmatherne-wk) [inline arelle/CntlrCmdLine.py:1095]: Does this need an assertion (and documentation) that this should only be used with traditional XBRL?

What happens if this runs on an inline doc? Does it remove the fact and value from the HTML? I guess a stretch goal could be to remove the fact tagging from iXBRL while preserving the rendered number (replace the XBRL element with its inner text), but I think exiting with an error is good enough until someone asks us to support iXBRL. iXBRL support has limited (no?) value while introducing many complexities (document sets, multiple targets, etc.)

Also, I think this should probably exit after saving the deduplicated instance instead of continuing on (log an error if the user tries to do other actions, validate, etc.) It appears that this puts the model into a strange state (we've modified the lxml tree to remove facts, but the model XBRL still includes them all).

Comment (aaroncameron-wk) [reply arelle/CntlrCmdLine.py]: Opted to just raise an error if this is run on any document that isn't a traditional XBRL instance. I do think iXBRL deduplication would be possible, but it makes sense to wait for demand.

I moved deduplication to run as late as possible since it does put the model into an undefined state. For what it's worth, I do think there's a way to write a new deduplicated XBRL document without manipulating the in-memory ModelXbrl (probably involves walking the lxml tree and writing every element that *isn't* a duplicate fact), but doing it this way allowed reuse of `ModelXbrl.saveInstance`. I'm willing to explore this approach if there's a consensus that it is worthwhile.

It seems that the only case where the model would potentially be used after the model is manipulated is if `keepOpen` is active or plugins that make use of `CntlrCmdLine.Filing.Validate` or `CntlrCmdLine.Filing.End` are active.

Comment (austinmatherne-wk) [reply arelle/CntlrCmdLine.py]: `CntlrCmdLine.Filing.End` and `CntlrCmdLine.Filing.Validate` are pretty common hooks. Can you move this after those hooks and raise an exception if `keepOpen` is used along with `saveDeduplicatedInstance`?

Comment (aaroncameron-wk) [reply arelle/CntlrCmdLine.py]: I've updated this to raise an error after saving if `keepOpen` is used.

The `ModelXbrl` is closed immediately after deduplication occurs currently (unless `keepOpen` is enabled), so it can't be moved any later. Looking into it more, it would appear that if either `CntlrCmdLine.Filing.End` and `CntlrCmdLine.Filing.Validate` are used by a plugin, they won't have access to an open `ModelXbrl` unless they're also using `keepOpen`, so I don't think moving this is necessary.

File: arelle/CntlrCmdLine.py
Comment (austinmatherne-wk) [inline arelle/CntlrCmdLine.py:1127]: I believe both this and the assert below should be `self.addToLog(..., level=logging.CRITICAL)` and set `success` to false.

File: arelle/ValidateDuplicateFacts.py
Comment (aaroncameron-wk) [reply arelle/ValidateDuplicateFacts.py:207]: This implementation (`consistent-pairs` deduplication) delivers the end result of having no consistent duplicates in the output. I think "removing all consistent duplicate facts" is a reasonable interpretation of "consistent deduplication". Even though it's a higher time-complexity implementation, it seems fair to provide this option for anyone expecting this deduplication result.

The other implementation (`consistent-sets` deduplication) may retain consistent duplicates in the output since it takes a hands-off approach when inconsistent duplicates are encountered. We could rename the arg to `consistent` to nudge users in that direction if they're not sure.

Comment (aaroncameron-wk) [reply arelle/ValidateDuplicateFacts.py:207]: The intent was that this loop did not iterate on the final group (the group of highest precision) since facts in that group will never be removed due to having a higher-precision duplicate. I see now that I incorrectly expected the index to be exclusive.

Comment (brettkail-wk) [reply arelle/ValidateDuplicateFacts.py]: > This implementation (consistent-pairs deduplication) delivers the end result of having no consistent duplicates in the output

The only difference between `consistent-sets` and `consistent-pairs` is when the report has inconsistent duplicates, which means the report is already of "questionable" validity.  If a user asks to deduplicate, they probably consider inconsistent duplicates to be a fatal error, so I'm trying to understand in what scenario they would care about deduplicating amongst the inconsistent duplicates.  Giving an option that is equivalent to `consistent-sets` when the report is good but runs slower when the report needs more changes seems odd to me.

> We could rename the arg to consistent to nudge users in that direction if they're not sure.

Renaming `consistent-sets` to the shorter `consistent` seems good to me.  I'm not sure about `consistent-pairs` since a1,a2,a3 and b1,b2,b3 with values a,b at three varying precision levels still reduces to a3,b3, so at a high-level it's not about "pairs", it's that it tries hard to remove consistent duplicates even if there are inconsistent duplicates.  Perhaps `complete`, `consistent`, and `inconsistent` are reasonable names for what they do (particularly if `inconsistent` is documented as being potentially slow)?

Anyway, I'll let you all decide.

**Review by derekgengenbacher-wf — APPROVED** [review]

File: arelle/CntlrCmdLine.py
Comment (austinmatherne-wk) [reply arelle/CntlrCmdLine.py]: Thanks. I missed that the models were closed before those hooks executed.

**Review by austinmatherne-wk — APPROVED** [review]

**Pull request merged**
